main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala

```diff
@@ -0,0 +1,352 @@
+
+        case Type.Cst(TypeConstructor.JvmMethod(method), _) =>
+          tracker.markProgress()
+          Type.getFlixType(method.getReturnType)
+
+        case Type.Cst(TypeConstructor.JvmField(field), _) =>
+          tracker.markProgress()
+          Type.getFlixType(field.getType)
+
+        case t => Type.JvmToType(t, loc)
+      }
+
+    case Type.JvmToEff(tpe, loc) =>
+      reduce(tpe) match {
+        case Type.Cst(TypeConstructor.JvmConstructor(constructor), _) =>
+          tracker.markProgress()
+          BaseEffects.getConstructorEffs(constructor, loc)
+
+        case Type.Cst(TypeConstructor.JvmMethod(method), _) =>
+          tracker.markProgress()
+          BaseEffects.getMethodEffs(method, loc)
+
+        case t => Type.JvmToType(t, loc)
+      }
+
+    case unresolved@Type.UnresolvedJvmType(member, loc) =>
+      member.map(reduce) match {
+        case JvmMember.JvmConstructor(clazz, tpes) =>
+          lookupConstructor(clazz, tpes) match {
+            case JavaConstructorResolution.Resolved(constructor) =>
+              tracker.markProgress()
+              Type.Cst(TypeConstructor.JvmConstructor(constructor), loc)
+            case _ => unresolved
+          }
+
+        case JvmMember.JvmField(tpe, name) =>
+          lookupField(tpe, name.name) match {
+            case JavaFieldResolution.Resolved(field) =>
+              tracker.markProgress()
+              Type.Cst(TypeConstructor.JvmField(field), loc)
+            case _ => unresolved
+          }
+
+        case JvmMember.JvmMethod(tpe, name, tpes) =>
+          lookupMethod(tpe, name.name, tpes) match {
+            case JavaMethodResolution.Resolved(method) =>
+              tracker.markProgress()
+              Type.Cst(TypeConstructor.JvmMethod(method), loc)
+            case _ => unresolved
+          }
+
+        case JvmMember.JvmStaticMethod(clazz, name, tpes) =>
+          lookupStaticMethod(clazz, name.name, tpes) match {
+            case JavaMethodResolution.Resolved(method) =>
+              tracker.markProgress()
+              Type.Cst(TypeConstructor.JvmMethod(method), loc)
+            case _ => unresolved
+          }
+      }
+  }
+
+  /** Tries to find a constructor of `clazz` that takes arguments of type `ts`. */
+  def lookupConstructor(clazz: Class[?], ts: List[Type]): JavaConstructorResolution = {
+    val typesAreKnown = ts.forall(isKnown)
+    if (!typesAreKnown) return JavaConstructorResolution.UnresolvedTypes
+
+    val tparams = ts.map(getJavaType)
+    val c = ConstructorUtils.getMatchingAccessibleConstructor(clazz, tparams *)
+
+    // Check if we found a matching constructor.
+    if (c != null) {
+      JavaConstructorResolution.Resolved(c)
+    } else {
+      JavaConstructorResolution.NotFound
+    }
+  }
+
+  /** Tries to find a method of `thisObj` that takes arguments of type `ts`. */
+  def lookupMethod(thisObj: Type, methodName: String, ts: List[Type]): JavaMethodResolution = {
```

Comment:
What methods should be public vs private?

Reply:
Probably almost all private